ocaml/testsuite/tests/typing-zero-alloc/signatures.ml

```diff
@@ -0,0 +1,821 @@
+end
+
+module M2_6 : S2_1 = struct
+  let[@zero_alloc assume strict never_returns_normally] f x = x
+end
+
+[%%expect{|
+module type S2_1 = sig val f : 'a -> 'a [@@zero_alloc] end
+module M2_1 : S2_1
+module M2_2 : S2_1
+module M2_3 : S2_1
+module M2_4 : S2_1
+module M2_5 : S2_1
+module M2_6 : S2_1
+|}]
+
+module type S2_2 = sig
+  val[@zero_alloc opt] f : 'a -> 'a
+end
+
+module M2_7 : S2_2 = struct
+  let[@zero_alloc] f x = x
+end
+
+module M2_8 : S2_2 = struct
+  let[@zero_alloc opt] f x = x
+end
+
+module M2_9 : S2_2 = struct
+  let[@zero_alloc assume] f x = x
+end
+
+module M2_10 : S2_2 = struct
+  let[@zero_alloc assume never_returns_normally] f x = x
+end
+
+module M2_11 : S2_2 = struct
+  let[@zero_alloc strict] f x = x
+end
+
+module M2_12 : S2_2 = struct
+  let[@zero_alloc strict opt] f x = x
+end
+
+module M2_13 : S2_2 = struct
+  let[@zero_alloc assume strict] f x = x
+end
+
+module M2_14 : S2_2 = struct
+  let[@zero_alloc assume strict never_returns_normally] f x = x
+end
+
+[%%expect{|
+module type S2_2 = sig val f : 'a -> 'a [@@zero_alloc opt] end
+module M2_7 : S2_2
+module M2_8 : S2_2
+module M2_9 : S2_2
+module M2_10 : S2_2
+module M2_11 : S2_2
+module M2_12 : S2_2
+module M2_13 : S2_2
+module M2_14 : S2_2
+|}]
+
+module type S2_3 = sig
+  val[@zero_alloc strict] f : 'a -> 'a
+end
+
+module M2_15 : S2_3 = struct
+  let[@zero_alloc strict] f x = x
+end
+
+module M2_16 : S2_3 = struct
+  let[@zero_alloc assume strict] f x = x
+end
+
+module M2_17 : S2_3 = struct
+  let[@zero_alloc assume strict never_returns_normally] f x = x
+end
+
+[%%expect{|
+module type S2_3 = sig val f : 'a -> 'a [@@zero_alloc strict] end
+module M2_15 : S2_3
+module M2_16 : S2_3
+module M2_17 : S2_3
+|}]
+
+module type S2_4 = sig
+  val[@zero_alloc strict opt] f : 'a -> 'a
+end
+
+module M2_18 : S2_4 = struct
+  let[@zero_alloc strict] f x = x
+end
+
+module M2_19 : S2_4 = struct
+  let[@zero_alloc strict opt] f x = x
+end
+
+module M2_20 : S2_4 = struct
+  let[@zero_alloc assume strict] f x = x
+end
+
+module M2_21 : S2_4 = struct
+  let[@zero_alloc assume strict never_returns_normally] f x = x
+end
+
+[%%expect{|
+module type S2_4 = sig val f : 'a -> 'a [@@zero_alloc strict opt] end
+module M2_18 : S2_4
+module M2_19 : S2_4
+module M2_20 : S2_4
+module M2_21 : S2_4
+|}]
+
+(*********************************)
+(* Test 3: disallowed inclusions *)
+
+module type S3_1 = sig
+  val[@zero_alloc] f : 'a -> 'a
+end
+
+module M3_1 : S3_1 = struct
+  let f x = x
+end
+
+[%%expect{|
+module type S3_1 = sig val f : 'a -> 'a [@@zero_alloc] end
+Lines 5-7, characters 21-3:
+5 | .....................struct
+6 |   let f x = x
+7 | end
+Error: Signature mismatch:
+       Modules do not match: sig val f : 'a -> 'a end is not included in S3_1
+       Values do not match:
+         val f : 'a -> 'a
+       is not included in
+         val f : 'a -> 'a [@@zero_alloc]
+       The former provides a weaker "zero_alloc" guarantee than the latter
+|}]
```

Comment:
:)

It was easy enough to add a hint for the case where the attribute is entirely absent in the implementation, so I've done that.  Good idea.

Because the module inclusion check happens somewhat distantly from the attribute parsing on a more abstract version of the information in the attribute, it would be harder to provide exactly the payload the user needs to write, or to give a hint in cases where the attribute is present but just wrong. But I think that's fine - the new hint improves the case most likely to send users to Rust.